node/src/types/block/past_finality_signatures.rs

```diff
@@ -0,0 +1,268 @@
+use std::collections::BTreeSet;
+
+use casper_types::{
+    bytesrepr::{self, Bytes, FromBytes, ToBytes},
+    PublicKey,
+};
+
+use datasize::DataSize;
+use itertools::Itertools;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use tracing::error;
+
+/// List of identifiers for finality signatures for a particular past block.
+///
+/// That past block height is current_height - rewards_lag, the latter being defined
+/// in the chainspec.
+///
+/// We need to wait for a few blocks to pass (`rewards_lag`) to store the finality
+/// signers because we need a bit of time to get the block finality.
+#[derive(
+    Clone,
+    DataSize,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Hash,
+    Serialize,
+    Deserialize,
+    JsonSchema,
+    Debug,
+    Default,
+)]
+pub struct PastFinalitySignatures(Vec<u8>);
+
+impl PastFinalitySignatures {
+    #[allow(dead_code)] //TODO remove it when the code is used in a next ticket
+    pub fn from_validator_set<'a>(
+        public_keys: &BTreeSet<PublicKey>,
+        all_validators: impl IntoIterator<Item = &'a PublicKey>,
+    ) -> Self {
+        // Take the validators list
+        // Replace the ones who signed with 1 and the ones who didn't with 0
+        // Pack everything into bytes
+        let result = Self::pack(
+            all_validators
+                .into_iter()
+                .map(|key| u8::from(public_keys.contains(key))),
+        );
+
+        let included_count: u32 = result.0.iter().map(|c| c.count_ones()).sum();
+        if included_count as usize != public_keys.len() {
+            error!(
+                included_count,
+                expected_count = public_keys.len(),
+                "error creating past finality signatures from validator set"
+            );
+        }
+
+        result
+    }
+
+    #[allow(dead_code)] //TODO remove it when the code is used in a next ticket
+    pub fn into_validator_set(
+        self,
+        all_validators: impl IntoIterator<Item = PublicKey>,
+    ) -> BTreeSet<PublicKey> {
+        self.unpack()
+            .zip(all_validators.into_iter())
+            .filter_map(|(active, validator)| (active != 0).then_some(validator))
+            .collect()
+    }
+
+    /// Packs the bits to bytes, to create a `PastFinalitySignature`
+    /// from an iterator of bits.
+    pub(crate) fn pack(bits: impl Iterator<Item = u8>) -> Self {
+        let inner = bits
+            .chunks(8)
+            .into_iter()
+            .map(|bits_chunk| {
+                bits_chunk
+                    .enumerate()
+                    .fold(0, |acc, (index, value)| acc | (value << (7 - index)))
```

Comment:
This should probably check if the bits are in fact 0/1 - or we should document that the check should be done somewhere else.

Reply:
How would this check be? Do we want to log an error and ignore the number? Do we want to return a result (unikely I assume)?

Reply:
That's a good question. I'd go with either returning a result, or just considering any non-zero number as 1. The latter might be simpler for now, especially since I don't think `JsonBlock`s are ever used as inputs, only outputs.

Reply:
Done